src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -237,9 +288,73 @@ def on_left_up(self, evt):
         self._selected_shape = self._get_shape(evt)
         if self._selected_shape:
             self._selected_shape.on_left_up(evt)
+            state = self._selected_shape.get_state()
+            if state:
+                shape_state = ShapeState(self._selected_shape, state, self._is_new_shape, False)
+                if not self._undo_stack or self._undo_stack[-1] != shape_state:
+                    self._undo_stack.append(shape_state)
+                    self._redo_stack.clear()  # Clear redo stack when a shape's state is saved
         else:
             WorldOverlay.on_left_up(self, evt)
 
+    def undo(self):
+        """Undo the last action."""
+        if self._undo_stack:
```

Comment:
You could make the code  a tiny simpler to read by adjusting this `if` to
```
if not self._undo_stack:
     logging.warning(...)
     return
```
Then, the rest of the code is one indentation less, and become the "normal code".